### PR TITLE
Semantic status codes copied from gRPC-Go

### PR DESCRIPTION
**Pull request by punya**
**Description:**
Add semantic status codes (for use in, e.g., structured logs to help diagnose failures in receivers).

**Link to tracking Issue:** #4156

_Please delete paragraphs that you did not use before submitting._

### CODE REVIEW

File: status/status.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
+/*
+ *
+ * Copyright 2014 gRPC authors.
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply status/status.go:3]: The reason is that I want to not have them associated with gRPC but with Google, because people always tell me "why do you use gRPC status" when not using gRPC :)))

[the remark this answers is not in the document]
Comment (punya) [reply status/status.go:3]: That makes sense, so now I clarify the origin of these codes in doc.go.

Unfortunately, the generated code in googleapis does other undesirable things like coupling tightly to protobuf concepts and the protobuf runtime, so I really don't think it's a good idea to use that form verbatim.

File: status/status.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+	// Unauthenticated indicates the request does not have valid
+	// authentication credentials for the operation.
+	Unauthenticated Code = 16
+
+	_maxCode = 17
+)
+
+var strToCode = map[string]Code{
+	`"OK"`: OK,
+	`"CANCELLED"`:/* [sic] */ Canceled,
```
Comment (djaglowski) [inline status/status.go:118]: I know we are inheriting this, but would it make sense to have a comment that explains the reason why it should remain misspelled? 

```suggestion
	`"CANCELLED"`:/* [sic] */ Canceled,
	`"CANCELED"`:             Canceled,
```
Would it make sense to additionally support the correct spelling?

**Review by bogdandrutu — COMMENTED** [review]
Last comment: what is the relationship between this and `consumererror`? Are they different? If not enough different they should live close to each other.

File: status/status.go (hunk 2 above)
Comment (bogdandrutu) [reply status/status.go:118]: :))

Comment (punya) [reply status/status.go:118]: I'll switch to the correct spelling. The mapping is bi-directional, so there's no way to support both spellings in a consistent fashion. This will break anybody's who's trying to use the gRPC codes as a drop-in replacement, but we aren't claiming compatibility anyway.

I still think it's worth keeping the codes intended for situations that we haven't encountered yet in the collector, because they reflect a pragmatic way to think about the problem. If we discard them, I suspect we'll reinvent them, and probably worse, when we need them in future.

**Comment by punya** [comment]
@bogdandrutu wrote:

> what is the relationship between this and `consumererror`? Are they different? If not enough different they should live close to each other.

`consumererror` provides:
* a way to indicate that an error is permanent
* a way to bundle a log or trace with an error
* a deprecated way to combine multiple errors into one
All these utilities manipulate Go's `error` objects in memory.

By contrast, the `status` package
* defines status codes (not necessarily error codes)
* is intended for externally visible behavior - e.g. generating collector logs that are easier to parse mechanically

To me, this makes them different enough that they should be in separate packages.

**Comment by punya** [comment]
cc @Aneurysm9 to weigh in on Bogdan's question above (where should these status codes live relative to `consumererror`).

**Comment by bogdandrutu** [comment]
@punya indeed you define the codes, but most likely you will want to use the codes in the returned "errors" and that's where the overlap will start correct?

**Comment by jsuereth** [comment]
@bogdandrutu Are you assuming that you'll always be returning an `err` with this status code when it exists?

I'm not sure I by that, especially if we have any kindof error handling/logic around that.  It's entirely possible you could log an invalid input with a status code, but not reject the entire input (i.e. you can log that ONE incoming metric is bad, while still returning the remaining ones).

I.e. I think "consumererror" and "status code"s are separate.  I agree there's a lot of times you'll use them together, but it's not a 1:1 relationship.

**Comment by bogdandrutu** [comment]
@jsuereth based on my experience in Google that's how they are used. Also remember golang added the idea of "composite" errors, so we can wrap any system/third-party error into a status error.

**Review by Aneurysm9 — APPROVED** [review]
I think I agree that these are distinct from errors and should not be tied to `consumererror`.  For instance, as has recently been discussed in the spec, not finding a resource is not necessarily an error from the server perspective.

**Comment by jsuereth** [comment]
@bogdandrutu I think this PR may be muddying the water on "RPC status code" vs. "event status code".   This is more akin to how those status codes mapped to GWSLogs or other structured-logging events where you may opt to supress failing the stack, but you still want to log the error and pull out semantic information.


The idea behind these status codes (and tying them to log entries) is more around structured logging, and something we'd commonly use when evaluating user/system behavior in logs.  If you want to have a live chat I can remind of you some Google systems where this is more common place.   

For the Otel Collector, it's not clear to me that a partial-receiver failure, where you'd log failure with this status code, should lead to the *entire* receiver to fail.   If you think of this in the context of scraper-receivers, then it's likely (e.g. host metrics) that there could be multiple things "collecting" and one failure shouldn't ruin the whole bunch.   However, attaching logs for these partial failures and allowing the system to process what it can is more useful than just outright failing.   There's a lot of systems that match this internal to Google as well.


TL;DR; gRPC status code as the *default* return value is fine, but it's not the only (or the primary use case) for what we want in the collector receivers.
